File: blasd/syrk.go

// Copyright (c) Harri Rautila, 2013

// This file is part of github.com/hrautila/gomas package. It is free software,
// distributed under the terms of GNU Lesser General Public License Version 3, or
// any later version. See the COPYING tile included in this archive.

package blasd

// #cgo CFLAGS: -O3 -march=native -fomit-frame-pointer -ffast-math -Iinc -I.
// #cgo LDFLAGS: -lm
// #include "inc/interfaces.h"
import  "C"
import "unsafe"
import "github.com/hrautila/cmat"
import "github.com/hrautila/gomas"

func syrk(Cc, Ac *cmat.FloatMatrix, alpha, beta float64, bits, P, S, E int, conf *gomas.Config) error {
    var Am, Cm C.mdata_t
    
    Am.md = (*C.double)(unsafe.Pointer(&Ac.Data()[0]))
    Am.step = C.int(Ac.Stride())
    Cm.md = (*C.double)(unsafe.Pointer(&Cc.Data()[0]))
    Cm.step = C.int(Cc.Stride())

    C.__d_rank_blk(
        (*C.mdata_t)(unsafe.Pointer(&Cm)), 
        (*C.mdata_t)(unsafe.Pointer(&Am)), 
        C.double(alpha), C.double(beta), C.int(bits),
        C.int(P), C.int(S), C.int(E),
        C.int(conf.KB), C.int(conf.NB), C.int(conf.MB))
    return nil
}

/*
 * UpdateSym performs symmetric rank-k update C = beta*C + alpha*A*A.T or
 * C = beta*C + alpha*A.T*A if gomas.TRANS bit is set.
 */
func UpdateSym(c, a *cmat.FloatMatrix, alpha, beta float64, bits int, confs... *gomas.Config) *gomas.Error {

    conf := gomas.DefaultConf()
    if len(confs) > 0 {
        conf = confs[0]
    }

    ok := true
    cr, cc := c.Size()
    ar, ac := a.Size()
    if cr*cc == 0 {
        return nil
    }
    P := ac
    E := cr
    if bits & gomas.TRANS != 0 && bits & gomas.TRANSA == 0 {
        bits |= gomas.TRANSA
    }
    switch {
    case bits & gomas.TRANSA != 0:
        ok = cr == cc && cr == ac 
        P = ar
    default:
        ok = cr == cc && cr == ar 
    }
    if !ok {
        return gomas.NewError(gomas.ESIZE, "UpdateSym")
    } 
    if conf.NProc == 1 || conf.WB <= 0 || E <= conf.WB {
        syrk(c, a, alpha, beta, bits, P, 0, E, conf)
        return nil
    }

    // parallelized
    var sbits int = 0
    wait := make(chan int, 4)
    nM, nN := blocking(E, E, conf.WB)
    nT := 0
    if bits & gomas.TRANS != 0 {
        sbits |= gomas.TRANSA
    } else {
        sbits |= gomas.TRANSB
    }
    if bits & gomas.LOWER != 0 {
        sbits |= gomas.LOWER
        for j := 0; j < nN; j++ {
            jS := blockIndex(j, nN, conf.WB, E)
            jL := blockIndex(j+1, nN, conf.WB, E)
            // update lower trapezoidal/triangular blocks
            task := func(q chan int) {
                updtrm(c, a, a, alpha, beta, sbits, P, jS, jL, jS, E, conf)
                //syrk(c, a, alpha, beta, bits, P, jS, jL, conf)
                q <- 1
            }
            conf.Sched.Schedule(gomas.NewTask(task, wait))
            nT += 1
        }
    } else {
        sbits |= gomas.UPPER
        for j := 0; j < nM; j++ {
            jS := blockIndex(j, nM, conf.WB, E)
            jL := blockIndex(j+1, nM, conf.WB, E)
            // update upper trapezoidal/triangular blocks
            task := func(q chan int) {
                updtrm(c, a, a, alpha, beta, sbits, P, jS, E, jS, jL, conf)
                //syrk(c, a, alpha, beta, bits, P, jS, jL, conf)
                q <- 1
            }
            conf.Sched.Schedule(gomas.NewTask(task, wait))
            nT += 1
        }
    }
    for nT > 0 {
        <- wait
        nT -= 1
    }
    return nil
}

// Local Variables:
// tab-width: 4
// indent-tabs-mode: nil
// End:

File: blasd/syrk.c

// Copyright (c) Harri Rautila, 2013

// This file is part of github.com/hrautila/gomas package. It is free software,
// distributed under the terms of GNU Lesser General Public License Version 3, or
// any later version. See the COPYING tile included in this archive.


#include "dtype.h"
#include "interfaces.h"
#include "mvec_nosimd.h"
#include "print.h"


void
__rank_diag(mdata_t *C, const mdata_t *A, const mdata_t *B, 
              DTYPE alpha, DTYPE beta,  int flags,  int P, int nC, cache_t *cache)
{
  register int i, incA, incB, trans;
  //mdata_t C0 = {C->md, C->step};
  mdata_t A0 = {A->md, A->step};
  mdata_t B0 = {B->md, B->step};

  incA = flags & GOMAS_TRANSA ? A->step : 1;
  incB = flags & GOMAS_TRANSA ? 1 : B->step;
  trans = flags & GOMAS_TRANSA ? GOMAS_TRANSA : GOMAS_TRANSB;

  //printf("__rank_diag: P=%d, nC=%d, incA=%d, incB=%d\n", P, nC, incA, incB);
  if (flags & GOMAS_UPPER) {
    for (i = 0; i < nC; i++) {
      // scale the target row with beta
      __vscale(C->md, C->step, beta, nC-i);
      // update one row of C  (nC-i columns, 1 row)
      __gemm_colwise_inner_no_scale(C, &A0, &B0, alpha, trans,
                                    P, nC-i, 1, cache); 
      // move along the diagonal to next row of C
      C->md += C->step + 1;
      // move A to next row
      A0.md += incA;
      // move B to next column
      B0.md += incA; 
    }
  } else {
    for (i = 0; i < nC; i++) {
      // scale the target row with beta
      __vscale(C->md, C->step, beta, i+1);
      // update one row of C  (nC-i columns, 1 row)
      __gemm_colwise_inner_no_scale(C, &A0, &B0, alpha, trans,
                                      P, i+1, 1, cache); 
      // move to next row of C
      C->md ++;
      // move A to next row
      A0.md += incA;
      //__blk_print(&C0, nC, nC, "C0", "%8.1e");
    }
  }
}

/*
 * Symmetric rank update
 *
 * upper
 *   C00 C01 C02    C00  C01 C02     A0  
 *    0  C11 C12 =   0   C11 C12  +  A1 * B0 B1 B2
 *    0   0  C22     0   0   C22     A2
 *
 * lower:
 *   C00  0   0    C00   0   0      A0  
 *   C10 C11  0 =  C10  C11  0   +  A1 * B0 B1 B2
 *   C20 C21 C22   C20  C21 C22     A2

 */
void __rank_blk(mdata_t *C, const mdata_t *A,
                DTYPE alpha, DTYPE beta,
                int flags,  int P, int S, int E, int KB, int NB, int MB)
{
  register int i, j, nI, nC;
  mdata_t Cd, Ad, Bd;
  mdata_t Acpy, Bcpy;
  cache_t cache;
  DTYPE Abuf[MAX_KB*MAX_MB], Bbuf[MAX_KB*MAX_NB] __attribute__((aligned(64)));

  if (E-S <= 0 || P <= 0)
    return;

  // restrict block sizes as data is copied to aligned buffers of predefined max sizes.
  if (NB > MAX_NB || NB <= 0) {
    NB = MAX_NB;
  }
  if (MB > MAX_MB || MB <= 0) {
    MB = MAX_MB;
  }
  if (KB > MAX_KB || KB <= 0) {
    KB = MAX_KB;
  }

  // set to zero in order avoid NaN values later
  memset(Abuf, 0, sizeof(Abuf));
  memset(Bbuf, 0, sizeof(Bbuf));

  Acpy = (mdata_t){Abuf, MAX_KB};
  Bcpy = (mdata_t){Bbuf, MAX_KB};
  cache = (cache_t){&Acpy, &Bcpy, KB, NB, MB};

  if ((flags & GOMAS_TRANSA) || (flags & GOMAS_TRANS)) {
    for (i = S; i < E; i += NB) {
      nI = E - i < NB ? E - i : NB;
    
      __subblock(&Cd, C, i, i);
      __subblock(&Ad, A, 0, i);

      // 1. update on diagonal
      __rank_diag(&Cd, &Ad, &Ad, alpha, beta, flags, P, nI, &cache);

      // 2. update block right of diagonal (UPPER) or left of diagonal (LOWER)
      __subblock(&Ad, A, 0, i);
      if (flags & GOMAS_LOWER) {
        __subblock(&Cd, C, i, 0);
        __subblock(&Bd, A, 0, S);
        nC = i;
      } else {
        __subblock(&Cd, C, i, i+nI);
        __subblock(&Bd, A, 0, i+nI);
        nC = E - i - nI;
      }

      __gemm_colwise_inner_scale_c(&Cd, &Ad, &Bd, alpha, beta, GOMAS_TRANSA,
                                   P, 0, nC, 0, nI, &cache); 
    }
  } else {
    for (i = S; i < E; i += NB) {
      nI = E - i < NB ? E - i : NB;
    
      // 1. update on diagonal
      __subblock(&Cd, C, i, i);
      __subblock(&Ad, A, i, 0);
      __rank_diag(&Cd, &Ad, &Ad, alpha, beta, flags, P, nI, &cache);

      // 2. update block right of diagonal (UPPER) or left of diagonal (LOWER)
      __subblock(&Ad, A, i, 0);
      if (flags & GOMAS_LOWER) {
        __subblock(&Cd, C, i, 0);
        __subblock(&Bd, A, S, 0);
        nC = i;
      } else {
        __subblock(&Cd, C, i,    i+nI);
        __subblock(&Bd, A, i+nI, 0);
        nC = E - i - nI;
      }

      __gemm_colwise_inner_scale_c(&Cd, &Ad, &Bd, alpha, beta, GOMAS_TRANSB,
                                   P, 0, nC, 0, nI, &cache);
    }
  }
}



// Local Variables:
// indent-tabs-mode: nil
// End:

